### ISSUE TITLE
Unsafe assignment of an `any` value when using module resolution: "Bundler"

### ISSUE THREAD

**Issue by kingston** [typescript-eslint/typescript-eslint#8101]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I am working on transitioning from moduleResolution: "Node" to moduleResolution: "Bundler" especially since Vite 5.0 uses Roll-Up 4 which requires Bundler. However, for some reason, ESLint Typescript with type checking will resolve any icon imported from the "react-icons" library as "any" resulting in a "Unsafe assignment of an `any` value" error.

This is strange since it compiles just fine with Typescript and resolves to the correct type according to VSCode. Only ESLint is complaining that it resolves to an "any" type. Therefore, it appears there is a discrepancy with how Typescript is resolving the type and ESLint Typescript is resolving the type when the module resolution is set to Bundler. When module resolution is Node, it works just fine.

I tried debugging why this was occurring but couldn't find anything obvious other than the module resolution behavior is somehow different between ESLint's typescript parser and Typescript compiler's parser.

Example code:
```
import { HiPlus, HiMinus } from 'react-icons/hi2';

...
const SomeIcon = count % 2 === 0 ? HiPlus : HiMinus;
...
```

### Reproduction Repository Link

https://github.com/kingston/react-icons-eslint-repro

### Repro Steps

1) clone the repo
2) `yarn install`
3) `yarn lint`

You should see an error in App.tsx

.../react-icons-eslint-repro/src/App.tsx
  9:9  error  Unsafe assignment of an `any` value  @typescript-eslint/no-unsafe-assignment

Changing moduleResolution in tsconfig.json to Node fixes this issue (though results in other errors in another file).

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.15.0` |
| `@typescript-eslint/parser`            | `6.15.0` |
| `TypeScript`                           | `5.2.2` |
| `ESLint`                               | `8.56.0` |
| `node`                                 | `20.10.0` |

**Comment by bradzacher** [comment]
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/641b495b-702f-4288-a062-351c646f30c9)

The lint rule is correct - there's a problem with their type declarations which means that their `IconType` type is `any`. If you turn off `skipLibCheck` and run `tsc` you'll see the error:

```
$ yarn tsc -p tsconfig.json --noEmit
node_modules/react-icons/hi2/index.d.ts:2:36 - error TS7016: Could not find a declaration file for module '../lib'. 'node_modules/react-icons/lib/cjs/index.js' implicitly has an 'any' type.

2 import { IconTree, IconType } from '../lib'
                                     ~~~~~~~~


Found 1 error in node_modules/react-icons/hi2/index.d.ts:2

**Comment by kingston** [comment]
Ah my bad. I didn't realize to disable skipLibCheck. Thank you very much. It's a react-icons issue and I'll see what can be done on that side.
